format date columns issue partialy solved

ARMSTRONGOPONDO opened the pull request:
@Ag-Derek
@lilyclements
@rdstern
added a format tab radio button the new column and format description  dialog #4282

Armstrongsoftwareeng committed:
format date columns issue partialy solved

lilyclements commented:
@ARMSTRONGOPONDO great. Is this ready for review or are you still working on it?

Armstrongsoftwareeng committed:
crop Definitions -Add tooltips

rdstern commented:
@ARMSTRONGOPONDO it is now 2 weeks since the @lilyclements question above.  You have been silent and I wonder what you are working on, if not this?
So I decided to look:
a) Developer error which needs to be fixed, when loading the dialog.
b) Here is the current dialog:

<img width="717" height="899" alt="image" src="https://github.com/user-attachments/assets/31c6ebee-9087-4b04-b97f-9c3f477e35d0" />

So I tried typing into the receiver, but can't get Ok to be enabled.  So I am unable to check if you have started on the R code.

This all seems remarkably little so far.  Maybe it was 30 minutes of work?

Please let us know when you have made some progress and reply before then to say it is not yet ready for testing.

Armstrongsoftwareeng committed:
the complete fix on the issue on format date

ARMSTRONGOPONDO reviewed (commented):
This pull request completely( fixes #4282 ) @Ag-Derek @lilyclements @rdstern 

<img width="482" height="605" alt="Screenshot 2025-07-23 084415" src="https://github.com/user-attachments/assets/2adeda20-c282-4664-8b3b-a4424458e0af" />


<img width="582" height="483" alt="Screenshot 2025-07-23 084259" src="https://github.com/user-attachments/assets/5d7312e0-0bad-4c82-a206-c0363ca99809" />

lilyclements reviewed (changes requested):
@ARMSTRONGOPONDO this is great! Really well done!! Working perfectly. The only changes are in the input text itself.

There are two small changes outlined below

Armstrongsoftwareeng committed:
added the help button and updated the examples to the one that was provided

Armstrongsoftwareeng committed:
moved the help button to the suggested position

rdstern commented:
@ARMSTRONGOPONDO and @lilyclements, here are some suggestions:
a) Please can you add a control at the bottom to be able to choose the name of the new variable.  This is in many dialogs, like the Transform dialog, see below:  

<img width="657" height="706" alt="image" src="https://github.com/user-attachments/assets/443414bb-9709-43af-9d15-c76a2988e558" />


It is the control at the bottom namely New Column Name, up to the Position button.

b) A smaller change, could you label your help button as `R Help`, and also mention it as R Help in the text below - so you say `Click 'R Help' to see the full list`
<img width="720" height="892" alt="image" src="https://github.com/user-attachments/assets/b12c841d-eb4c-44bb-8c24-7a12add9f815" />

c) Could you delete - well move really - the examples section from here.  That will makie that control neater, because it won't need the scroll bar:

<img width="624" height="218" alt="image" src="https://github.com/user-attachments/assets/d8efe369-a5f9-4204-b273-8c1b3c5010ad" />

d)  Could you please add all those examples at the top of the pull down, shown in your current dialog above.  You could include the brackets too, e.g. `%b %-d       (Jan 5)`, though only the first part, e.g. `%b %-d` will be transferred to the R function.  Alternatively the `Jan 5` could be displayed as a tooltip, when you hover over that entry?  (I am suggesting they be added at the top, because I think they will be used more often than the single parts of the date you give now.)

lilyclements commented:
@ARMSTRONGOPONDO can you send up your most recent changes so I can have a look into this error?

Armstrongsoftwareeng committed:
Add date formatting control and enhance UI elements

- Introduced `ucrFormatNew` control for specifying new column names in `dlgUseDate`.
- Updated help button text to "R Help" for clarity.
- Improved `ComboBoxnewformat` with specific date format examples and tooltips.
- Enhanced `TestOKEnabled` method to check `ucrFormatNew` completeness.

ARMSTRONGOPONDO commented:
@lilyclements here are the implemented changes that ive pushed today ( I had a problem with my laptop)
<img width="482" height="609" alt="image" src="https://github.com/user-attachments/assets/40cf58a6-217e-4cc2-ac50-8cf35ab7ae1c" />

This pull request introduces significant improvements to the date formatting functionality in the `dlgUseDate` dialog, enhancing both usability and flexibility for users. The changes add a new "format column" mode, allow users to construct custom date formats with helpful tooltips and examples, and provide direct access to R's date-time format help. Additionally, tooltips in the `dlgPICSACrops` dialog have been expanded for clarity.

**Date Formatting Enhancements:**

* Added a new "format column" mode to `dlgUseDate`, allowing users to specify custom date formats and create new columns with formatted dates. This includes new radio buttons to toggle between using an existing column and formatting with a custom pattern, and updates to the logic that controls which UI elements are visible and which R parameters are set. [[1]](diffhunk://#diff-3fdd9d07b58d456e89d95f41b1f5f67c8654618471982b3a61a238062c797adeR201-R209) [[2]](diffhunk://#diff-3fdd9d07b58d456e89d95f41b1f5f67c8654618471982b3a61a238062c797adeR288-R345)
* Populated the date format dropdown (`ComboBoxnewformat`) with common R date format patterns and individual format codes, and provided dynamic tooltips to guide users in selecting or constructing their own formats.
* Added a help button (`cmdHelp`) that opens the R documentation for `strptime`, making it easier for users to find all available date-time format codes. [[1]](diffhunk://#diff-3fdd9d07b58d456e89d95f41b1f5f67c8654618471982b3a61a238062c797adeR51-R58) [[2]](diffhunk://#diff-3fdd9d07b58d456e89d95f41b1f5f67c8654618471982b3a61a238062c797adeL245-R414)
* Updated the resource file (`DlgUseDate.resx`) to include a clear description of available date format codes for users.

**Usability Improvements in Crop Dialog:**

* Expanded tooltips in the `dlgPICSACrops` dialog to include examples for entering comma-separated sequences, improving user understanding of input formats for planting dates, crop durations, and water amounts. [[1]](diffhunk://#diff-297f8a7b0c4ea5c4ed386c04035947a5ee95c37721855d3866a09df57620c7d5L126-R130) [[2]](diffhunk://#diff-297f8a7b0c4ea5c4ed386c04035947a5ee95c37721855d3866a09df57620c7d5L135-R154)

lilyclements reviewed (commented):
@ARMSTRONGOPONDO good news! You've done it - and you've done too much if anythign!

You are running, say:

```r
format_date <- data_book$split_date(data_name="ghana", col_name="date", format_date=TRUE, format_string="%b %d", new_col_name="format_date", s_start_month=1)
data_book$add_columns_to_data(data_name="ghana", col_name="format_date", col_data=format_date, before=FALSE, adjacent_column="date")
```

But we just want to run:

```r
data_book$split_date(data_name="ghana", col_name="date", format_date=TRUE, format_string="%b %d", new_col_name="format_date", s_start_month=1)
```

That is, we want to not have the assign to, and not to have the `$add_columns_to_data`.

I've outlined below how that can be achieved (my 1-6 points below). 
I've also given other small bits and pieces to help improve this code. 

Nice one!

ARMSTRONGOPONDO committed:
Implemented the suggested changes

ARMSTRONGOPONDO committed:
merging the branch to match the master so that I can get the translations changes from master

ARMSTRONGOPONDO committed:
Add new keys to JSON and adjust "Outfilled" entry

Added three new key-value pairs: "PICSA Trend Graph", "Use ",
and "New Format" to support new functionality. Re-added the
"Outfilled" key, likely to correct formatting or placement.

lilyclements reviewed (dismissed):
@ARMSTRONGOPONDO really nice! Works well and looks good too. @Vitalis95 can you review on the R side?

lilyclements committed:
Merge branch 'master' into format_date

ARMSTRONGOPONDO committed:
resolved the translations faiuling

ARMSTRONGOPONDO commented on instat/translations/en/r_instat_not_menus.json:
```diff
@@ -6848,4 +6848,8 @@
   "PICSA Trend Graph": "PICSA Trend Graph",
   "Use ": "Use ",
   "New Format": "New Format"
```
You missed a comma on this line, which is why the translation  was failing after resolving the conflict. I have already corrected it.

Vitalis95 commented on instat/DlgUseDate.vb:
```diff
@@ -172,39 +218,69 @@ Public Class dlgUseDate
+            ucr.bUpdateRCodeFromControl = True
+        Next
+
+        rdoUseColumn.Checked = True
```
@ARMSTRONGOPONDO , one small suggestion, instead of hardcoding this, could you  check how `ucrPnlSelectData` is implemented in `dlgCluster` ?

ARMSTRONGOPONDO committed:
Removed the hardcoded
 ucr.bUpdateRCodeFromContro

ARMSTRONGOPONDO committed:
Merge branch 'master' into format_date

ARMSTRONGOPONDO committed:
Update r_instat_not_menus.json

resolving the conflict

ARMSTRONGOPONDO committed:
resolving the translations build error

lilyclements reviewed (commented):
@ARMSTRONGOPONDO on the json, it just looks like we can use the current unmerged json from the master branch, and just need to add in 

```
 "Use ": "Use ",
 "New Format": "New Format",
```

Does that sound right to you?

ARMSTRONGOPONDO committed:
Merge branch 'master' into format_date

lilyclements committed:
Merge branch 'master' into format_date

ARMSTRONGOPONDO committed:
retrigur the buld

ARMSTRONGOPONDO committed:
Merge branch 'master' of https://github.com/ARMSTRONGOPONDO/R-Instat into format_date

lilyclements reviewed (commented):
@ARMSTRONGOPONDO amazing, well done! Works really nicely too, and a good addition to the system!

I noticed that when I press reset, the ucrSave and the Format dropdown boxes do not reset. That's the only thing I could break. I tried pretty hard to break it too, so, nice job. Can you fix the reset, and then let me know and I'll approve :)

ARMSTRONGOPONDO committed:
Introduced `bReset` in the reset handler to ensure proper reset behavior.

lilyclements reviewed (dismissed):
@ARMSTRONGOPONDO amazing - looks great and works well! @Vitalis95 can you review the VB code?
